Relative local weights

ShirmanDenis opened the pull request.

Commit:
add relative weights modifier

Commit:
refactoring

Commit:
refactoring 2

Commit:
fix logging

Commit:
add more tests

Commit:
add interfaces for testing

Commit:
fix tests

kungasc commented on Vostok.ClusterClient.Core/Misc/KeyValuePairExtensions.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Vostok.Clusterclient.Core.Misc
+{
+    public static class KeyValuePairExtensions
```
```suggestion
    internal static class KeyValuePairExtensions
```

Commit:
CR comments

Commit:
CR fixes[1]
# make KeyValuePairExtensions internal
# fix StatisticBucket not all fields reads was throw interlocked
# make WeighingHelper static class

Commit:
CR fixes[2]
# fix concurrency bug in Weights

Commit:
CR fixes[3]
# normalize weights relative to Min and Max from IWeightedReplicaOrderingBuilder
# remove lock on updating weights
# apply weight regeneration for replicas

Commit:
CR fixes[4]
# rename StatisticSnapshot to ClusterStatistic
# StatsiticHistory now remove outdated statistic on update
# some refactoring and renames

Commit:
small refactoring

KiskachiMaria committed:
CR 2

Commit:
renames

Commit:
refactoring

Commit:
rename

Commit:
Merge branch 'master' into relative_local_weights

Commit:
add weights normalizer
# normalize only new weights

Commit:
RelativeWeightModifier: fix weight normalization when max weight is positive infinity

Commit:
Merge branch 'master' into relative_local_weights

Commit:
calculate relative weights by statuses when low rps

Commit:
RelativeWeightSettings: change defaults

Commit:
move ObtainGlobalValue() to new interface instead extending IReplicaStorageProvider

Commit:
RelativeWeightModifier: change default global storage provider to per process
